src/libstore/store-uri.hh

```diff
@@ -0,0 +1,100 @@
+#pragma once
+///@file
+
+#include <variant>
+
+#include "types.hh"
+
+namespace nix {
+
+/**
+ * A parsed Store URI (URI is a slight misnomer...), parsed but not yet
+ * resolved to a specific instance and query parms validated.
+ *
+ * Supported values are:
+ *
+ * - ‘local’: The Nix store in /nix/store and database in
+ *   /nix/var/nix/db, accessed directly.
+ *
+ * - ‘daemon’: The Nix store accessed via a Unix domain socket
+ *   connection to nix-daemon.
+ *
+ * - ‘unix://<path>’: The Nix store accessed via a Unix domain socket
+ *   connection to nix-daemon, with the socket located at <path>.
+ *
+ * - ‘auto’ or ‘’: Equivalent to ‘local’ or ‘daemon’ depending on
+ *   whether the user has write access to the local Nix
+ *   store/database.
+ *
+ * - ‘file://<path>’: A binary cache stored in <path>.
+ *
+ * - ‘https://<path>’: A binary cache accessed via HTTP.
+ *
+ * - ‘s3://<path>’: A writable binary cache stored on Amazon's Simple
+ *   Storage Service.
+ *
+ * - ‘ssh://[user@]<host>’: A remote Nix store accessed by running
+ *   ‘nix-store --serve’ via SSH.
+ *
+ * You can pass parameters to the store type by appending
+ * ‘?key=value&key=value&...’ to the URI.
+ */
+struct StoreURI {
+    using Params = std::map<std::string, std::string>;
+
+    /**
+     * Special keyword `` or `auto`
+     */
+    struct Auto {
+        inline auto operator <=> (const Auto & rhs) const = default;
+    };
+
+    /**
+     * Special keyword `local`
+     */
+    struct Local {
+        inline auto operator <=> (const Local & rhs) const = default;
+    };
+
+    /**
+     * Special keyword `daemon`
+     */
+    struct Daemon {
+        inline auto operator <=> (const Daemon & rhs) const = default;
+    };
+
+    /**
+     * General case, a regular `scheme://authority` URL.
+     */
+    struct Generic {
+        std::string scheme;
+        std::string authority;
+
+        auto operator <=> (const Generic & rhs) const = default;
+    };
+
+    typedef std::variant<
+        Auto,
+        Local,
+        Daemon,
+        Generic
+    > Variant;
```

Comment:
Since we're going through an intermediate representation, can't we just get rid of all that, and just make the non-uri things be aliases to their proper uri version? So `local` would be an alias for `local://`, `daemon` for `daemon://`, and `auto` for a new `auto://` `StoreFactory` that returns the right store.

Reply:
I would prefer this too. No harm is done, and we flatten the sum type into the scheme. That's a lot more bearable in C++.
The apparent downside is that it's less typeful, but the scheme was never meant to be a closed sum type, making the set of modeled constructors arbitrary. That doesn't seem helpful.

Reply:
So `local://` doesn't actually exist, which gets in the way of that, but maybe the simple answer is that it should.

Reply:
Oh and `auto` is dynamically resolved via IO, so it also has to stay separate.

Reply:
I'll at least try getting rid of `daemon`.

Reply:
Oh and `unix://` isn't a substitute for `daemon` either.
```shell-session
 $ nix doctor --store 'daemon'
[PASS] PATH contains only one nix version.
[PASS] All profiles are gcroots.
[PASS] Client protocol matches store protocol.
 $ nix doctor --store 'unix://'
[PASS] PATH contains only one nix version.
[PASS] All profiles are gcroots.
error: cannot connect to socket at '': Connection refused
```